pkg/network/ebpf/c/tracer/stats.h

```diff
@@ -362,4 +362,52 @@ static __always_inline int handle_tcp_recv(u64 pid_tgid, struct sock *skp, int r
     return handle_message(&t, 0, recv, CONN_DIRECTION_UNKNOWN, packets_out, packets_in, PACKET_COUNT_ABSOLUTE, skp);
 }
 
+__maybe_unused static __always_inline bool tcp_failed_connections_enabled() {
+    __u64 val = 0;
+    LOAD_CONSTANT("tcp_failed_connections_enabled", val);
+    return val > 0;
+}
+
+// handle_tcp_failure handles TCP connection failures on the socket pointer and adds them to the connection tuple
+// returns an integer to the caller indicating if there was a failure or not
+static __always_inline bool handle_tcp_failure(struct sock *sk, conn_tuple_t *t) {
+    if (!tcp_failed_connections_enabled()) {
+        return false;
+    }
+    int err = 0;
+    BPF_CORE_READ_INTO(&err, sk, sk_err);
+
+    switch (err) {
+    case 0:
+    return false; // no error
+    case TCP_CONN_FAILED_RESET:
+    case TCP_CONN_FAILED_TIMEOUT:
+    case TCP_CONN_FAILED_REFUSED: {
+        tcp_stats_t stats = { .failure_reason = err };
+        update_tcp_stats(t, stats);
+        return true;
+    }
+    case TCP_CONN_FAILED_HOSTDOWN:
+        increment_telemetry_count(tcp_failure_ehostdown);
+        return false;
+    case TCP_CONN_FAILED_HOSTUNREACH:
+        increment_telemetry_count(tcp_failure_ehostunreach);
+        return false;
+    case TCP_CONN_FAILED_NETDOWN:
+        increment_telemetry_count(tcp_failure_enetdown);
+        return false;
+    case TCP_CONN_FAILED_NETUNREACH:
+        increment_telemetry_count(tcp_failure_enetunreach);
+        return false;
+    case TCP_CONN_FAILED_NETRESET:
+        increment_telemetry_count(tcp_failure_enetreset);
+        return false;
+    case TCP_CONN_FAILED_ABORTED:
+        increment_telemetry_count(tcp_failure_econnaborted);
+        return false;
+    }
+    increment_telemetry_count(unsupported_tcp_failures);
```

Comment:
Would it be possible to do a map instead? Like error code -> count

Reply:
we can create a map and then scrape it in userspace when metrics are collected yes